File: 748.shortest-completing-word.cpp
/*
 * @lc app=leetcode id=748 lang=cpp
 *
 * [748] Shortest Completing Word
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  string shortestCompletingWord(string licensePlate, vector<string>& words) {
    int need[26];
    memset(need, 0, sizeof(need));
    for (const auto c : licensePlate) {
      if (isalpha(c)) {
        if (c >= 'A' && c <= 'Z') {
          need[c - 'A']++;
        } else {
          need[c - 'a']++;
        }
      }
    }

    // Use not std::sort but std::stable_sort, to preserve order for same length
    std::stable_sort(words.begin(), words.end(), [](const auto& lhs, const auto& rhs) {
      return lhs.length() < rhs.length();
    });

    for (const auto& word : words) {
      int copy[26];
      memcpy(copy, need, sizeof(copy));
      if (isCompleting(copy, word)) {
        return word;
      }
    }
    return "";
  }

 private:
  bool isCompleting(int need[], const string& word) {
    for (const auto c : word) {
      need[c - 'a']--;
    }
    for (int i = 0; i < 26; i++) {
      if (need[i] > 0) {
        return false;
      }
    }
    return true;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<string> words;
  words = {"step", "steps", "stripe", "stepple"};
  ASSERT_EQ(s.shortestCompletingWord("1s3 PSt", words), "steps");
  words = {"looks", "pest", "stew", "show"};
  ASSERT_EQ(s.shortestCompletingWord("1s3 456", words), "pest");
  words = {"suggest",   "letter", "of",      "husband", "easy",
           "education", "drug",   "prevent", "writer",  "old"};
  ASSERT_EQ(s.shortestCompletingWord("Ah71752", words), "husband");
  words = {"enough", "these",  "play",  "wide", "wonder",
           "box",    "arrive", "money", "tax",  "thus"};
  ASSERT_EQ(s.shortestCompletingWord("OgEu755", words), "enough");
  words = {"claim", "consumer", "student", "camera",  "public",
           "never", "wonder",   "simple",  "thought", "use"};
  ASSERT_EQ(s.shortestCompletingWord("iMSlpe4", words), "simple");
  words = {"bad", "and"};
  ASSERT_EQ(s.shortestCompletingWord("Ad1234", words), "bad");
  return 0;
}

File: 748.shortest-completing-word.rs
/*
 * @lc app=leetcode id=748 lang=rust
 *
 * [748] Shortest Completing Word
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn shortest_completing_word(license_plate: String, words: Vec<String>) -> String {
        let mut count = [0; 26];
        license_plate
            .chars()
            .filter_map(|c| {
                if c.is_alphabetic() {
                    Some(c.to_ascii_lowercase())
                } else {
                    None
                }
            })
            .for_each(|c| count[((c as u8) - b'a') as usize] += 1);

        let mut short_word = String::new();
        for word in words.into_iter() {
            let mut copy = [0; 26];
            word.as_bytes()
                .iter()
                .for_each(|&b| copy[(b - b'a') as usize] += 1);

            if count.iter().zip(copy.iter()).all(|(l, w)| *w >= *l)
                && (short_word.is_empty() || short_word.len() > word.len())
            {
                short_word = word;
            }
        }
        short_word
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::shortest_completing_word(
            "1s3 PSt".to_string(),
            vec![
                "step".to_string(),
                "steps".to_string(),
                "stripe".to_string(),
                "stepple".to_string()
            ]
        ),
        "steps".to_string()
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::shortest_completing_word(
            "1s3 456".to_string(),
            vec![
                "looks".to_string(),
                "pest".to_string(),
                "stew".to_string(),
                "show".to_string()
            ]
        ),
        "pest".to_string()
    );
}
